sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -511,6 +511,20 @@ class SparkSession private(
     createDataset(data.asScala.toSeq)
   }
 
+  /**
+   * Creates a [[Dataset]] from an RDD of spark.sql.catalyst.InternalRow. This method allows
+   * the caller to create externally the InternalRow set, as we as define the schema externally.
+   *
+   * @since 3.3.0
+   */
+  def createDataset(data: RDD[InternalRow], schema: StructType): DataFrame = {
```

Comment:
Hello @HyukjinKwon, I'm trying to connect the dots, and map your suggestion with the set goals.
I agree, internal rows are by principle internal and generated off the conversion between the external RDD types and the internal types. The main intention with this change, as mentioned on the PR description:

1 - allow additional flexibility when creating a typed DataSet, and corresponding rows outside Spark. The first change on StructType helps, because avoids unnecessary loops when building the rows.

2 - Improve overall performance with the external building of Spark rows. By passing the InternalRows in the dataset creation, we're bypassing the conversion to Catalyst rows and improvement the general performance of the driver.
So, assuming this not a path we want to pursuit - creating internalRows outside - we still need to create a dataframe based on the exposed Row type (at least that is how we need to do it in our project).
For that we can just used the existing method createDataFrame(rowRDD: RDD[Row], schema: StructType): DataFrame 

The above suggestion is assuming there is already a dataframe from which we would retrieve the Spark plan with the internal rows. What is exactly the intention with creating a DF from another DF if there isn't any change on data and schema? Apologies if it's an obvious answer, and I didn't get there yet. I'm a Spark contributor newbie :)

Reply:
@HyukjinKwon - should I go ahead and remove this method for the time being, so we can move forward with this PR? Or is there an opportunity to improve elsewhere? Please let me know your thoughts.
Happy to brainstorm further!

Reply:
@HyukjinKwon, if I don't hear back from you during the next couple of days, I'm going to remove this code, and just keep other change, so we can move forward this PR. Thanks!

Reply:
I was thinking about making it as a developer API, and users would be able to do:

```scala
val rdd: RDD[InternalRow] = ...
val attributtes = schema.fields.map(f => AttributeReference(f.name, f.dataType, f.nullable, f.metadata)())
Dataset.ofRows(spark, org.apache.spark.sql.execution.LogicalRDD(attributtes, rdd)(spark))
```

But I am not sure if this is the best approach. Probably we should better raise this discussion into Spark dev mailing (dev@spark.apache.org) list, and discuss further.

Reply:
Thanks for coming back. I've sent an email to the dev list. Could you please review it and add any additional comments to this new thread? Thanks!

Reply:
@HyukjinKwon  I've sent an email to dev@spark.apache.org, as well as dev-subscribe@spark.apache.org and I've got no response. I can find the subject on the mailing lists http://mail-archives.apache.org/mod_mbox/spark-dev/202111.mbox/browser, neither I'm receiving any emails from them. Could you please point me to what I could be doing wrong?

Reply:
@AmplabJenkins  anyway you could help me to make progress?